docs: RETURNS STATUS AS ROW (#4656)
yugabyte/yugabyte-db: docs/content/latest/api/ycql/batch.md

---
title: BATCH requests [YCQL]
headerTitle: BATCH
linkTitle: BATCH
summary: Execute multiple DML in 1 request
description: Use batch to update multiple rows in 1 request.
menu:
  latest:
    parent: api-cassandra
    weight: 19991
aliases:

>   - /latest/api/ycql/batch2

Reviewer: why `batch2`?

Author: It's no longer needed because docs have been updated.

isTocNested: true
showAsideToc: true
---

Batch operations let you send multiple operations in a single RPC call to the database. The larger the batch size, 
the higher the latency for the entire batch. Although the latency for the entire batch of operations is higher than the latency of any single operation, 
the throughput of the batch of operations is much higher.

> ## Example in JAVA

Reviewer 2: ```suggestion
## Example in Java
```

> In order to perform a batch insert operation in Java, first create a BatchStatement object. 
> Next add the desired number of prepared and bound insert statements to it. 
> Finally, execute the batch object. This is shown below.

Reviewer 2: ```suggestion
To perform a batch insert operation in Java:

1.  Create a BatchStatement object. 
2. Add the desired number of prepared and bound insert statements to it. 
3. Execute the batch object. 
```

```java
// Create a batch statement object.
BatchStatement batch = new BatchStatement();

// Create a prepared statement object to add to the batch.

> PreparedStatement insert = client.prepare("INSERT INTO table (k, v) VALUES (?, ?);");

Reviewer: don't need the semicolon at the end.

// Bind values to the prepared statement and add them to the batch.
for (...) {
  batch.add(insert.bind( ... <values for bind variables> ... ));
}

// Execute the batch operation.
ResultSet resultSet = client.execute(batch);

```
## Example in Python using RETURNS AS STATUS

An example using Python client and RETURNS AS STATUS clause:

> ```cassandra

Reviewer: Is it cassandra or cassandraql? A lot of the other snippets just use `sql` because there used to be no highlighting declared for cql.

ycqlsh> create keyspace if not exists yb_demo;
ycqlsh> CREATE TABLE if not exists yb_demo.test_rs_batch(h int, r bigint, v1 int, v2 varchar, primary key (h, r));
ycqlsh> INSERT INTO yb_demo.test_rs_batch(h,r,v1,v2) VALUES (1,1,1,'a');
ycqlsh> INSERT INTO yb_demo.test_rs_batch(h,r,v2) VALUES (3,3,'b');
ycqlsh> select * from yb_demo.test_rs_batch;

 h | r | v1   | v2
---+---+------+----
 1 | 1 |    1 |  a
 3 | 3 | null |  b

(2 rows)
```

Getting status from DML operations in Python:

```python
from cassandra.cluster import Cluster, BatchStatement

# Create a cluster and a session
cluster = Cluster(['127.0.0.1'])
session = cluster.connect()

# Create a batch statement object.
b = BatchStatement()

# Add multiple queries

> b.add(f"INSERT INTO test_rs_batch(h, r, v1, v2) VALUES (1, 1, 1 ,'a') RETURNS STATUS AS ROW;")

Reviewer: you should not need the semicolons at the end.

b.add(f"UPDATE test_rs_batch SET v2='z' WHERE h=3 AND r=3 IF v2='z'  RETURNS STATUS AS ROW;")
b.add(f"DELETE FROM test_rs_batch WHERE h=2 AND r=2 IF EXISTS RETURNS STATUS AS ROW;")

# Execute the batch operation.
result = session.execute(b, trace=True)

# Print status for each DML operation
for row in result:
    print(row)
```

The output generated is:

```python
Row(applied=True, message=None, h=None, r=None, v1=None, v2=None)
Row(applied=False, message=None, h=3, r=3, v1=None, v2='b')
Row(applied=False, message=None, h=None, r=None, v1=None, v2=None)
```

## Row Status

> When executing a batch in YCQL, the protocol just allows returning one error/return status.

Reviewer 2: ```suggestion
When executing a batch in YCQL, the protocol just allows returning one error or return status.
```
Avoid "/".

> So, if one statement fails with an error or, for conditional DMLs, some are not applied because of failing the IF condition, 
> the driver/application cannot accurately identify the relevant statement(s), it will just receive one general error/return-status for the whole batch.

Reviewer 2: ```suggestion
If one statement fails with an error or for conditional DMLs, some are not applied because of failing the IF condition, the driver or application cannot accurately identify the relevant statements, it will just receive one general error or return-status for the batch.
```

> Therefore, it is not trivial/possible for an application to react to such failures appropriately (e.g. retry, abort, 
> change some parameters, for either entire batch or just the relevant statements etc).
> 

Reviewer 2: ```suggestion
Therefore, it is not possible for an application to react to such failures appropriately (for example, retry, abort, 
and change some parameters for either the entire batch or just the relevant statements).

```

> The `RETURNS STATUS AS ROW` feature addresses this limitation.

Reviewer 2: ```suggestion
You can address this limitation by using the `RETURNS STATUS AS ROW`.
```

> If used, the (write) statement will return it's status (whether applied, unapplied, or errored-out with a message) as a regular CQL 
> row that the application can inspect and decide what to do.

Reviewer 2: ```suggestion
If used, the write statement will return its status (whether applied, unapplied, or errored-out with a message) as a regular CQL row that the application can inspect and decide what to do.
```

> In the case of a batch (which is its intended use-case), it is required that either none or all statements use `RETURNS STATUS AS ROW`.

Reviewer 2: ```suggestion
For a batch, it is required that either none or all statements use `RETURNS STATUS AS ROW`.
```

> This is so that when executing n statements in a batch with `RETURN STATUS AS ROW` it will always return n rows (in the same order as the statements) 

Reviewer 2: ```suggestion
When executing `n` statements in a batch with `RETURN STATUS AS ROW`, `n` rows are returned, in the same order as the statements. 
```

> This is so that when executing n statements in a batch with `RETURN STATUS AS ROW` it will always return n rows (in the same order as the statements) 
> and the app can easily inspect the result.

Reviewer 2: ```suggestion
When executing `n` statements in a batch with `RETURN STATUS AS ROW`, `n` rows are returned, in the same order as the statements and the application can easily inspect the result.
```

> Similarly, it is required that `RETURN STATUS AS ROW` is used for batches containing conditional DMLs.

Reviewer 2: ```suggestion
For batches containing conditional DMLs, `RETURN STATUS AS ROW` must be used.
```

> This is because, for conditional DMLs (which are normally not allowed in batches), any subset of them could fail due to 
> their `IF` condition so only returning rows for them would make it impossible to identify which ones actually failed.
> 
> To distinguish between the two not-applied cases (error vs condition is false) there is an (error) `message` column in 
> the return row that will be null for not-applied and filled-in for errors.
> 
> Conversely, there will be one column for each table column which will be `null` for errors but filled-in for not-applied 
> (justifying the decision to not apply).

Reviewer 2: ```suggestion
For conditional DMLs (not normally allowed in batches), any subset of them could fail due to 
their `IF` condition and thus returning rows only for them makes it impossible to identify which ones actually failed.

To distinguish between the two not-applied cases (error vs condition is false), there is an error `message` column in the return row that will be null for not-applied and filled-in for errors.

Conversely, there will be one column for each table column, which will be `null` for errors but filled-in for not-applied (justifying the decision to not apply).
```

For instance:

1. Set up a simple table:
```cassandraql

> cqlsh:sample> CREATE TABLE test(h INT, r INT, v LIST<INT>, PRIMARY KEY(h,r)) WITH transactions={'enabled': true};

Reviewer: does the copy button work fine for these cases?
Does it not also copy the output? Also the highlighting does not seem to work.

Reviewer 2: Use `sql` instead of `cassandraql`. AFAIK, the syntax identifiers that generate the "COPY" button include: `postgresql`, `sh`, and `sql`.

cqlsh:sample> INSERT INTO test(h,r,v) VALUES (1,1,[1,2]);
```

2. Unapplied update when `IF` condition is false:
```cassandraql

> cqlsh:sample> UPDATE test SET v[2] = 4 WHERE h = 1 AND r = 1 IF v[1] = 3 RETURNS STATUS AS ROW;

Reviewer: does the copy button work fine for these cases?
Does it not also copy the output? Also the highlighting does not seem to work.
Same for the other cases, check that all Copy's on the table extracts (only) the relevant CQL text.
